man/libvdeplug.3
.\"* libvdeplug: a network namespace as a user library
.\" Copyright (C) 2017 Renzo Davoli. University of Bologna. <renzo@cs.unibo.it>
.\" 
.\" This library is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Lesser General Public
.\" License as published by the Free Software Foundation; either
.\" version 2.1 of the License, or (at your option) any later version.
.\" 
.\" This library is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Lesser General Public License for more details.
.\" 
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with this library; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
	
.TH libvdeplug 3 2017-05-21 "VirtualSquare" "Linux Programmer's Manual"
.SH NAME
vde_open, vde_recv, vde_send, vde_send, vde_recv, vde_close \- connect clients to a VDE (virtual distributed ethernet) network
.SH SYNOPSIS
.B #include <libvdeplug.h>
.br
.BI "VDECONN *vde_open(char *vnl,char *descr, struct vde_open_args *open_args);"
.br
.BI "ssize_t vde_recv(VDECONN *conn,void *buf,size_t len,int flags);"
.br
.BI "ssize_t vde_send(VDECONN *conn,const void *buf,size_t len,int flags);"
.br
.BI "int vde_datafd(VDECONN *conn);"
.br
.BI "int vde_ctlfd(VDECONN *conn);"
.br
.BI "int vde_close(VDECONN *conn);"
.sp
These functions are provided by libvdeplug. Link with \fI-lvdeplug\fR.
.SH DESCRIPTION
Libvdeplug is the library used by clients to join a VDE network. It is a generic library which can
use several different implementations for virtual distributed networks. In fact the Virtual Network Locator (VNL) \fIvnl\fR 
has the syntax \fImodule\fB://\fIspecific_address\fR.
e.g. \fBvde:///home/user/myswitch\fR or \fBvxvde://239.0.0.1\fR.
The default module is \fBlibvdeplug_vde(1)\fR (connection to a \fBvde_switch(1)\fR)
and can be omitted. e.g. \fB/home/user/myswitch\fR means
\fBvde:///home/user/myswitch\fR

\fBvde_open\fR opens a VDE connection: \fIvnl\fR is the vde network identifier as described here above
and \fIdescr\fR is a description of this connection (it can be useful to identify this connection in
the port list of a switch). \fRopen_args\fR has been left for backwards compatibility but can be safely
set to NULL as all the options previosly defined as fields of \fIstruct vde_open_args\fR are now 
parameters of the VDE modules (can be defined using a module specific syntax in VNL).

\fBvde_recv\fR receives a packet from the VDE connection.

\fBvde_send\fR sends a packet to the VDE connection.

\fBvde_datafd\fR returns the data file descriptor of the connection: this descriptor can be used in 
\fBpoll(2)\fR or \fBselect(2)\fR (or similar) system calls. When this descriptor signals data available for reading
\fBvde_recv\fR will not block.

\fBvde_ctlfd\fR returns the control file descriptor of the connection. Not all the modules support control
descritors, but when it is valid descriptor (greater or equal than 0) can be used to check if the remote 
endpoint (switch or peer) hanged up. No data can be sent or received using the control file descriptor, 
it returns an end-of-file condition when the connection get closed (zero length on reading).

\fBvde_close\fR closes a vde connections.

.SH RETURN VALUE
\fBvde_open\fR returns the descriptor of the VDE connection which is used as a parameter of all the other functions.
NULL is returned in case of error.

\fBvde_recv\fR returns the length of the packet received. It should never be less then 14 bytes (the length of
		an ethernet header). It returns -1 in case of error, it may return 1 to notify that a packet has been
received but it must be dropped (e.g. the sender was not allowed to send that packet).

\fBvde_send\fR returns the number of bytes sent. It returns -1 in case of error.

\fBvde_datafd\fR and \fBvde_ctlfd\fR returns a file descriptor and  -1 in case of error.

\fBvde_datafd\fR returns 0 in case of success and  -1 in case of error.

.SH NOTICE
Virtual Distributed Ethernet is not related in any way with
www.vde.com ("Verband der Elektrotechnik, Elektronik und Informationstechnik"
i.e. the German "Association for Electrical, Electronic & Information
Technologies").

.SH SEE ALSO
\fBvde_plug\fR(1),
\fBvde_switch\fR(1)

.SH BUGS
Bug reports should be addressed to <info@virtualsquare.org>
.SH AUTHOR
Renzo Davoli <renzo@cs.unibo.it>


debian/changelog
vdeplug4 (0.1-1) unstable; urgency=medium

  * Initial release (Closes: #0000)

 -- Mattia Biondi <mattia.biondi3@studio.unibo.it>  Sat, 14 Sep 2019 10:14:53 +0200


libvdeplug4/libvdeplug.c
/*
 * libvdeplug - A library to connect to a VDE Switch.
 * Copyright (C) 2013-2016 Renzo Davoli, University of Bologna
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation version 2.1 of the License, or (at
 * your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301, USA
 */

#define _GNU_SOURCE
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <limits.h>
#include <dlfcn.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <libvdeplug.h>
#include "libvdeplug_mod.h"

/* Per-User standard switch definition */
/* This will be prefixed by getenv("HOME") */
/* it can be a symbolic link to the switch dir */
#define STDSWITCH "/.vdeplug/default"
#define OLDSTDSWITCH "/.vde2/default.switch"

#define STD_LIB_PATH "/lib:/usr/lib"
#define LIBVDEPLUG "libvdeplug_"
#define LIBVDEPLUGIN "/vdeplug/" LIBVDEPLUG

#define DEFAULT_MODULE "vde"
uint64_t __vde_version_tag = 4;

struct vde_open_parms {
	char *vde_args;
	char *descr;
	int interface_version;
	struct vde_open_args *open_args;
};

static VDECONN *vde_open_module_lib(char *filename, struct vde_open_parms *parms) {
	void *handle=dlopen(filename, RTLD_NOW | RTLD_DEEPBIND);

	//printf("TRY %s %s\n", filename, parms->vde_args);
	if (handle) {
		struct vdeplug_module *module=dlsym(handle, "vdeplug_ops");
		if (module) {
			VDECONN *rv = module->vde_open_real(parms->vde_args, parms->descr, parms->interface_version, parms->open_args);
			if (rv) {
				rv->handle = handle;
				rv->module = module;
				return rv;
			}
		} else
			errno = EPROTONOSUPPORT;
		dlclose(handle);
	} else
		errno = EPROTONOSUPPORT;
	return NULL;
}

static VDECONN *vde_open_samedir(char *modname, char *subdir, struct vde_open_parms *parms) {
	char path[PATH_MAX];
	Dl_info info;
	if (dladdr(&__vde_version_tag, &info) != 0) {
		const char *libpath = info.dli_fname;
		char *slash = strrchr(libpath, '/');
		if (slash) {
			int len = slash - libpath;
			snprintf(path, PATH_MAX, "%*.*s%s%s.so",
					len, len, libpath, subdir, modname);
			return vde_open_module_lib(path, parms);
		}
	}
	return errno = EPROTONOSUPPORT, NULL;
}

static VDECONN *vde_open_dirlist(char *modname, char *dirlist, char *subdir, struct vde_open_parms *parms) {
	char path[PATH_MAX];
	VDECONN *retval;
	if (dirlist) {
		while (*dirlist != 0) {
			int len = strchrnul(dirlist, ':') - dirlist;
			snprintf(path, PATH_MAX, "%*.*s%s%s.so",
					len, len, dirlist, subdir, modname);
			retval = vde_open_module_lib(path, parms);
			if (retval != NULL || errno != EPROTONOSUPPORT)
				return retval;
			dirlist += len + (dirlist[len] == ':');
		}
	}
	return errno = EPROTONOSUPPORT, NULL;
}

static VDECONN *vde_open_default(char *modname, struct vde_open_parms *parms) {
	char path[PATH_MAX];
	snprintf(path, PATH_MAX, LIBVDEPLUG "%s.so", modname);
	return vde_open_module_lib(path, parms);
}

static VDECONN *vde_open_module(char *modname, struct vde_open_parms *parms) {
	char *vdeplugin_path = getenv("VDEPLUGIN_PATH");
	// creata a copy of parms->vde_args
	size_t vde_args_len = strlen(parms->vde_args);
	char vde_args_copy[vde_args_len + 1];
	strcpy(vde_args_copy, parms->vde_args);
	parms->vde_args = vde_args_copy;

	// if VDEPLUGIN_PATH exists it defines the search path for plugins
	if (vdeplugin_path)
		return vde_open_dirlist(modname, vdeplugin_path, "/" LIBVDEPLUG, parms);
	else {
		VDECONN *retval;
		retval = vde_open_samedir(modname, LIBVDEPLUGIN, parms);
		if (retval != NULL || errno != EPROTONOSUPPORT) return retval;
		retval = vde_open_samedir(modname, "/" LIBVDEPLUG, parms);
		if (retval != NULL || errno != EPROTONOSUPPORT) return retval;
		retval = vde_open_dirlist(modname, getenv("LD_LIBRARY_PATH"), LIBVDEPLUGIN, parms);
		if (retval != NULL || errno != EPROTONOSUPPORT) return retval;
		retval = vde_open_dirlist(modname, STD_LIB_PATH, LIBVDEPLUGIN, parms);
		if (retval != NULL || errno != EPROTONOSUPPORT) return retval;
		return vde_open_default(modname, parms);
	}
}

static void set_newdescr(const char *descr, char *newdescr, size_t newdescrlen) {
	char *ssh_client = getenv("SSH_CLIENT");
	int pid = getpid();
	uid_t uid = geteuid();
	FILE *f = fmemopen(newdescr, newdescrlen, "w");

	fprintf(f, "%s ", descr);

	fprintf(f, "user=%d", uid);
	fprintf(f, " pid=%d", pid);

	if (ssh_client) {
		int len = strchrnul(ssh_client,' ') - ssh_client;
		fprintf(f, " SSH=%*.*s", len, len, ssh_client);
	}

	fclose(f);
	newdescr[newdescrlen - 1] = 0;
	//printf("%s %d\n", newdescr, bufsize);
}

VDECONN *vde_open_real(char *vde_url, char *descr, int interface_version,
		struct vde_open_args *open_args)
{
	char std_vde_url[PATH_MAX];
	struct stat statbuf;
	char newdescr[MAXDESCR];
	char *tag;
	struct vde_open_parms parms = {
		.descr = newdescr,
		.interface_version = interface_version,
		.open_args = open_args
	};

	if (vde_url == NULL)
		vde_url = "";

	tag = strstr(vde_url, "://");

	set_newdescr(descr, newdescr, MAXDESCR);

	if (tag == NULL) {
		if (*vde_url == '\0') {
			char *homedir = getenv("HOME");
			if (homedir) {
				snprintf(std_vde_url, PATH_MAX, "%s%s", homedir, STDSWITCH);
				if (lstat(std_vde_url, &statbuf) >= 0)
					vde_url = std_vde_url;
				else {
					snprintf(std_vde_url, PATH_MAX, "%s%s", homedir, OLDSTDSWITCH);
					if (lstat(std_vde_url, &statbuf) >= 0)
						vde_url = std_vde_url;
				}
			}
		}
		if (stat(vde_url, &statbuf) >= 0) {
			if (S_ISREG(statbuf.st_mode)) {
				FILE *f=fopen(vde_url,"r");
				if (f != NULL) {
					if (fgets(std_vde_url, PATH_MAX, f) != NULL) {
						std_vde_url[strlen(std_vde_url) - 1] = 0;
						vde_url = std_vde_url;
					}
					tag = strstr(vde_url, "://");
					fclose(f);
				}
			}
		}
	}
	if (vde_url == NULL || tag == NULL) {
		parms.vde_args = vde_url;
		return vde_open_module(DEFAULT_MODULE, &parms);
	} else {
		int modlen=tag - vde_url;
		char modname[modlen + 1];
		snprintf(modname, modlen + 1, "%*.*s", modlen, modlen, vde_url);
		parms.vde_args = vde_url + (modlen + 3);
		return vde_open_module(modname, &parms);
	}
}

ssize_t vde_recv(VDECONN *conn,void *buf,size_t len,int flags)
{
	if (__builtin_expect(conn!=0,1))
		return conn->module->vde_recv(conn,buf,len,flags);
	else {
		errno=EBADF;
		return -1;
	}
}

ssize_t vde_send(VDECONN *conn,const void *buf,size_t len,int flags)
{
	if (__builtin_expect(conn!=0,1))
		return conn->module->vde_send(conn,buf,len,flags);
	else {
		errno=EBADF;
		return -1;
	}
}

int vde_datafd(VDECONN *conn)
{
	if (__builtin_expect(conn!=0,1))
		return conn->module->vde_datafd(conn);
	else {
		errno=EBADF;
		return -1;
	}
}

int vde_ctlfd(VDECONN *conn)
{
	if (__builtin_expect(conn!=0,1))
		return conn->module->vde_ctlfd(conn);
	else {
		errno=EBADF;
		return -1;
	}
}

int vde_close(VDECONN *conn)
{
	if (__builtin_expect(conn!=0,1)) {
		void *handle=conn->handle;
		/* vde_close frees the struct conn */
		int rv=conn->module->vde_close(conn);
		if (rv==0) dlclose(handle);
		return rv;
	} else {
		errno=EBADF;
		return -1;
	}
}
